### ISSUE TITLE
dynamic.option function

### ISSUE THREAD

**Issue by CrowdHailer** [gleam-lang/suggestions#78]
This comes up a lot when decoding json.

```rust
  assert Ok(name) = dynamic.field(data, "name")
  let name = case dynamic.string(name) {
    Ok(name) -> Some(name)
    // Assumes it's nil
    _ -> None
  }
```
Something like the following might be helpful

```rust
assert Ok(name) = dynamic.field(data, "name")
assert Ok(name) = dynamic.option(name, dynamic.string)
```

**Comment by lpil** [comment]
What would the `None` values be? The atoms `undefined`, `none` and `nil`?

**Comment by CrowdHailer** [comment]
Yes, I realized there was a bit of choice here, erlang does a few times use `[]` for nil a few times as well.
The canonical set you listed above does cover 95% of cases, however I am not sure if you would be endlessly getting asked about the other None cases.

If this is considered a good idea, does a dynamic.result also make sense?

**Comment by lpil** [comment]
I was thinking about one for Result, but I couldn't work out exactly how it would behave. It would be nice to coerse the Erlang result types of `ok | {ok, A} | error | {error, B}`

**Comment by CrowdHailer** [comment]
With `OK` (the elixir library) I never tackled anything but the tuple varients,

Nil doesn't really work, because you'll need to return `Result(Dynamic, Dynamic)` and the function won't work if returning Error(Nil) in the `error` case or Error(Dynamic) in the `{error, B}` case except at runtime

**Comment by lpil** [comment]
Sorry, yes, that's what I meant. It would be a dynamic nil

**Comment by CrowdHailer** [comment]
So the only `dynamic_option` function I am using in reality is one that takes the atom `null` as a nil value (not one we mentioned above) because it is returned from the pgo library, and I guess using SQL semantics. 
I think either this function would have to take a list of values it considers None (which would be quite messy) or it's not likely to be very useful

**Comment by lpil** [comment]
Let's do this one :)
